### ISSUE TITLE
API: Menu item enablement

### ISSUE THREAD

**Issue by bpasero** [microsoft/vscode#21131]
This used to be like that in the old world I think.

**Comment by joaomoreno** [comment]
@jrieken Any thoughts on contributed menu items enablement?

**Comment by jrieken** [comment]
Yeah, there is no story for enablement yet. For now you make the when-clause more specific such that the menu item doesn't show. One idea could be to add an enablement-clause when defining a command. It could use the when-syntax but define enablement of a command.

**Comment by joaomoreno** [comment]
Renamed for clarity.

**Comment by garyking** [comment]
It would be great to have this, so that the Open Changes button can be greyed out when there are no changes for the current file.

**Comment by jrieken** [comment]
cc @rebornix

**Comment by jrieken** [comment]
most work actually already happened via https://github.com/Microsoft/vscode/issues/41103, exposing this to extensions is the missing piece

**Comment by jrieken** [comment]
Idea would be to allow for a `precondition` with the command definition. That would allow to define the condition that must be fulfilled so that this command can be run, the UX is enablement/disablement.

```json
"contributes": {
    "commands": [
      {
        "command": "someCommand",
        "title": "My Some Command",
        "precondition": "myContext = 42",
        "category": "Something"
      },
```

**Comment by jrieken** [comment]
This happened via: https://github.com/microsoft/vscode/pull/74017

**Comment by jrieken** [comment]
@rebornix While revising #74017 I missed that enablement isn't a menu property but a command property. Commands should be enable/disabled independent of their UX location. I have pushed a change for this. After discussing we have also concluded that `enablement` is the better name for this. Sample a command that is only enabled when having a selection in the editor

```json
{
  "command": "extension.sayHello1",
  "title": "Hello Enablement",
  "enablement": "editorHasSelection"
}
```
